parser.js

```diff
@@ -0,0 +1,199 @@
+
+const blockHashTokens = ['hash']
+const blockHeightTokens = ['height']
+const blockTimeMedianTokens = ['mediantime']
+const blockTimeReceivedTokens = ['received_time']
+const blockTimeTokens = ['time']
+const blockTransactionTokens = ['tx', 'txids']
+const blockTransactionTotalTokens = ['n_tx']
+
+const transactionFeeTokens = ['fee', 'fees']
+const transactionInputIndexTokens = ['output_index', 'vout']
+const transactionInputTokens = [ 'inputs', 'vin']
+const transactionInputTxidTokens = ['txid', 'prev_hash']
+const transactionOutputIndexTokens = ['n']
+const transactionOutputTokens = ['outputs', 'vout']
+const transactionTimeConfirmedTokens = ['confirmed']
+const transactionTimeReceivedTokens = ['received', 'time']
+const transactionTxidTokens = ['txid', 'hash']
+const transactionValueTokens = ['value', 'output_value', 'total']
+
+const populateInputs = (transaction, inputs) => {
+  for (let i = 0; i < inputs.length; i++) {
+    let index = undefined
+    let txid = undefined
+    let value = undefined
+
+    for (const key of Object.keys(inputs[i])) {
+      if (transactionInputIndexTokens.includes(key)) {
+        index = inputs[i][key]
+        continue
+      }
+
+      if (transactionInputTxidTokens.includes(key)) {
+        txid = inputs[i][key]
+        continue
+      }
+
+      if (transactionValueTokens.includes(key)) {
+        value = inputs[i][key]
+        continue
+      }
+
+      if (txid && index && value) {
+        break
+      }
+    }
+
+    transaction.inputs.push(new TransactionInput(txid, index, value))
+  }
+}
+
+const populateOutputs = (transaction, outputs) => {
+  for (let i = 0; i < outputs.length; i++) {
+    let index = undefined
+    let value = undefined
+
+    for (const key of Object.keys(outputs[i])) {
+      if (transactionOutputIndexTokens.includes(key)) {
+        index = outputs[i][key]
+        continue
+      }
+
+      if (transactionValueTokens.includes(key)) {
+        value = outputs[i][key]
+        continue
+      }
+
+      if (index && value) {
+        break
+      }
+    }
+
+    transaction.outputs.push(new TransactionOutput(index, value))
+  }
+}
+
+// scannerless transaction parser which checks tx attributes against matching string tokens (does not transform data)
+const txParser = (rawTx) => {
+  if (typeof(rawTx) === 'string') {
+    rawTx = JSON.parse(rawTx)
+  }
+
+  let transaction = new Transaction()
+
+  for (const key of Object.keys(rawTx)) {
+    if (Array.isArray(rawTx[key])) {
+      if (transactionInputTokens.includes(key)) {
+        populateInputs(transaction, rawTx[key])
+        continue
+      }
+
+      if (transactionOutputTokens.includes(key)) {
+        populateOutputs(transaction, rawTx[key])
+        continue
+      }
+    } else {
+      if (transactionTxidTokens.includes(key)) {
+        transaction.txid = rawTx[key]
+        continue
+      }
+
+      if (transactionFeeTokens.includes(key)) {
```

Comment:
repeating this same logic over and over seems redundant, but the `continue` statement available inside the for loop provides a nice performance boost since we know if we have a match, we *don't* want to bother checking other tokens (this could change) - should test behaviour of ignoring IDE warning and just placing unsyntactic continue inside separate function which encapsulates this repeated logic